docs/design/values.md

```diff
@@ -0,0 +1,917 @@
+the restrictions on values.
+
+If a further escape hatch is needed, this kind of fundamental weakening of the
+semantic model would be a good case for some syntactic marker like Rust's
+`unsafe`, although rather than a region, it would seem better to tie it directly
+to the operation in question. For example:
+
+```carbon
+class S {
+  fn ValueMemberFunction[me: Self]();
+  fn AddrMemberFunction[addr me: const Self*]();
+}
+
+fn F(s_value: S) {
+  // This is fine.
+  s_value.ValueMemberFunction();
+
+  // This requires an unsafe marker in the syntax.
+  s_value.unsafe MutableMemberFunction();
+}
+```
+
+## Initializing expressions
+
+Storage in Carbon is initialized using _initializing expressions_. Their
+evaluation produces an initialized object in the storage, although that object
+may still be _unformed_.
+
+**Future work:** More details on initialization and unformed objects should be
+added to the design from the proposal
+[#257](https://github.com/carbon-language/carbon-lang/pull/257), see
+[#1993](https://github.com/carbon-language/carbon-lang/issues/1993). When added,
+it should be linked from here for the details on the initialization semantics
+specifically.
+
+The simplest form of initializing expressions are value or durable reference
+expressions that are converted into an initializing expression. Value
+expressions are written directly into the storage to form a new object.
+Reference expressions have the object they refer to copied into a new object in
+the provided storage.
+
+**Future work:** The design should be expanded to fully cover how copying is
+managed and linked to from here.
+
+The first place where an initializing expression is _required_ is to satisfy
+[_variable patterns_](#binding-patterns-and-local-variables-with-let-and-var).
+These require the expression they match to be an initializing expression for the
+storage they create. The simplest example is the expression after the `=` in a
+local `var` declaration.
+
+The next place where a Carbon expression requires an initializing expression is
+the expression operand to `return` statements. We expand more completely on how
+return statements interact with expressions, values, objects, and storage
+[below](#function-calls-and-returns).
+
+The last path that requires forming an initializing expression in Carbon is when
+attempting to convert a non-reference expression into an ephemeral reference
+expression, where temporary storage is materialized and provided to an
+initializing expression. The resulting object can then form the ephemeral
+reference expression.
+
+### Function calls and returns
+
+Functions with a return type and return statement in Carbon are called
+_returning functions_. Calls to these functions are modeled directly as
+initializing expressions -- they require storage as an input and when evaluated
+cause that storage to be initialized with an object. This means that when a
```

Comment:
This sounds like it means that `F(G())` can't be implemented by moving the return-value register to the first-argument register -- the return value of `G()` has to be materialized in memory. Is that accurate?

Reply:
By "notionally" do you mean those operations can only be omitted under as-if? i.e. if the object passed from `G` to `F` has a user-defined type, we have to actually perform two copy/move operations unless the compiler can prove they have no observable side effects? If so, it'd be nice if we could avoid constraining ourselves that way.

Of course, we can always add rules later to permit eliding those operations in some cases. I do wonder if there's another way of describing the semantics where the notional-but-elided operations don't occur in the first place (as there turned out to be for RVO in C++), but it probably doesn't make sense to further delay this proposal to try to find it.